fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreTest.java

```diff
@@ -3272,4 +3272,68 @@ public IndexState needRebuildIndex(Index index, long recordCount, boolean indexO
             assertTrue(recordStore.getRecordStoreState().getStoreHeader().getOmitUnsplitRecordSuffix());
         }
     }
+
+    @Test
+    public void testIndexStateFormatVersionUpgrade() {
+        final RecordMetaDataBuilder builder = RecordMetaData.newBuilder().setRecords(TestNoIndexesProto.getDescriptor());
+        try (FDBRecordContext context = openContext()) {
+            recordStore = FDBRecordStore.newBuilder()
+                    .setContext(context)
+                    .setMetaDataProvider(builder)
+                    .setKeySpacePath(path)
+                    .setFormatVersion(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1)
+                    .create();
+            assertEquals(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1, recordStore.getFormatVersion());
+            // Save 200 records without any indexes.
+            for (int i = 0; i < 200; i++) {
+                TestNoIndexesProto.MySimpleRecord record = TestNoIndexesProto.MySimpleRecord.newBuilder()
+                        .setRecNo(i).setNumValue(i + 1000).build();
+                recordStore.saveRecord(record);
+            }
+            commit(context);
+        }
+
+        // Add a new index named "index" on rec_no.
+        final Index index = new Index("index", "rec_no");
+        index.setSubspaceKey(1);
+        builder.addIndex("MySimpleRecord", index);
+        try (FDBRecordContext context = openContext()) {
+            recordStore = FDBRecordStore.newBuilder()
+                    .setContext(context)
+                    .setMetaDataProvider(builder)
+                    .setKeySpacePath(path)
+                    .setFormatVersion(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1)
+                    .createOrOpen();
```

Comment:
```suggestion
                    .open();
```

Reply:
It also might make sense to have a single builder where the key space path and format versions are set, and then it can be used with different contexts. That's just food for thought, though.